src/dist/replication/common/fs_manager.h

```diff
@@ -50,23 +50,36 @@ struct dir_node
     std::string full_dir;
     int64_t disk_capacity_mb;
     int64_t disk_available_mb;
-    int64_t disk_available_ratio;
+    int disk_available_ratio;
+    // disk_density = disk_available_ratio - total_available_ratio, it's used to evaluate whether a
+    // disk is balanced
+    int disk_density;
     std::map<app_id, std::set<gpid>> holding_replicas;
+    std::map<app_id, std::set<gpid>> holding_primary_replicas;
+    std::map<app_id, std::set<gpid>> holding_secondary_replicas;
 
 public:
-    dir_node(const std::string &tag_, const std::string &dir_)
+    dir_node(const std::string &tag_,
+             const std::string &dir_,
+             int64_t disk_capacity_mb_ = 0,
+             int64_t disk_available_mb_ = 0,
+             int disk_available_ratio_ = 0,
+             int disk_density_ = 0)
         : tag(tag_),
           full_dir(dir_),
-          disk_capacity_mb(0),
-          disk_available_mb(0),
-          disk_available_ratio(0)
+          disk_capacity_mb(disk_capacity_mb_),
+          disk_available_mb(disk_available_mb_),
+          disk_available_ratio(disk_available_ratio_),
+          disk_density(disk_density_)
     {
     }
     unsigned replicas_count(app_id id) const;
     unsigned replicas_count() const;
     bool has(const dsn::gpid &pid) const;
     unsigned remove(const dsn::gpid &pid);
     void update_disk_stat();
+
+    void set_disk_density(int density) { disk_density = density; }
```

Comment:
Is this function only used for function `compute_disk_density`?

Reply:
Yes

Reply:
If this function is only used for `compute_disk_density`, and this function is simple, there is no need to write a function for only one sentence.